clients/src/main/java/org/apache/kafka/clients/admin/ListGroupsOptions.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.admin;
+
+import org.apache.kafka.common.GroupType;
+import org.apache.kafka.common.annotation.InterfaceStability;
+
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * Options for {@link Admin#listGroups()}.
+ * <p>
+ * The API of this class is evolving, see {@link Admin} for details.
+ */
+@InterfaceStability.Evolving
+public class ListGroupsOptions extends AbstractOptions<ListGroupsOptions> {
+
+    private Set<GroupType> types = Collections.emptySet();
+
+    /**
+     * If types is set, only groups of these types will be returned by listGroups().
+     * Otherwise, all groups are returned.
+     */
+    public ListGroupsOptions withTypes(Set<GroupType> types) {
```

Comment:
Should we use String instead of `GroupType` to allow older clients to list 'future' groups? For example, older clients that only support Consumer and Classic wouldn't be able to list Share groups, as they don't have `GroupType.Share`.

Additionally, the returned `ConsumerGroupListing` should use a String as well.

@AndrewJSchofield @dajac WDYT?

Reply:
I prefer to use an enum in order to stay consistent with the rest of our public api and because using strings is error prone. If users want to use new features, they will likely upgrade their clients anyway.

Reply:
> I prefer to use an enum in order to stay consistent with the rest of our public api and because using strings is error prone. If users want to use new features, they will likely upgrade their clients anyway.

Okay, I also prefer enums and agree that strings can be error-prone, but I'd like to discuss this further. I assumed `ListGroup` was intended as a generic method capable of displaying all past and future group types. This also means that users would need to upgrade their admin tools to correctly parse GroupType, even if the RPC version remains unchanged.

Reply:
I prefer enums, but I am not 100% convinced with how this works at the moment. We have separate enums for the different group states, but I wonder whether we'd just be better off with a single `GroupState` enum that is the union of all of the existing groups. In practice, `ConsumerGroupState` contains all of the states of the other groups already.

This unlocks another possibility which is `ListGroupsOptions.withStates` which I consider worth having.

So, I suggest:
* Introduce `GroupState`
* Introduce `ListGroupsOptions.withStates`
* Work out what to do with values beyond the current enum so that an admin client has a chance to cope with new states or group types

I'm happy to take this on and do some experimentation if that's OK.

Reply:
> I prefer enums, but I am not 100% convinced with how this works at the moment. We have separate enums for the different group states, but I wonder whether we'd just be better off with a single GroupState enum that is the union of all of the existing groups. In practice, ConsumerGroupState contains all of the states of the other groups already.

I planned to use `state` as an example to respond to @dajac's argument. :smile: 

> This unlocks another possibility which is ListGroupsOptions.withStates which I consider worth having.

If the enum can cover all possible (including future) values, it would be ideal to use an enum in the query. As @dajac mentioned, an enum is preferable to a string in this case.

> Work out what to do with values beyond the current enum so that an admin client has a chance to cope with new states or group types

This wouldn't be an issue if we accept that users need to update their clients to use new features. Alternatively, we could keep using enums as query filters since users can omit the filters to query 'new' groups with an 'old' client. The real issue is that users will only see 'unknown' in the response, even if the RPC response carries the correct value. A workaround is to add a method to ListGroupsResult to return the 'raw value' for those "unknown" field.

Reply:
Java enums are really inflexible.

Here is my suggestion.
1) Create a new `GroupState` enum which is basically a copy of `ConsumerGroupState` but which applies to all group types. Deprecate `ConsumerGroupState`.
1) Add `ListGroupsOptions.inStates(Set<GroupState>)`. This means that `Admin.listGroups` is able to do everything that `Admin.listConsumerGroups` and `Admin.listShareGroups` can do.
1) `ConsumerGroupState` is referred to in 3 places: `ConsumerGroupListing` (deprecated by KIP-1043), `ListConsumerGroupsOptions` (also deprecated by KIP-1043), and `ConsumerGroupDescription`.
1) Add `ConsumerGroupDescription.groupState()` which returns `GroupState`. Deprecate `ConsumerGroupDescription.state()` which returns `ConsumerGroupState`.
1) Replace `ShareGroupState` with `GroupState`. Replace `ShareGroupDescription.state()` with `ShareGroupDesription.groupState()`.
1) Add `ConsumerGroupDescription.rawState()` which returns `String` so that a client can distinguish obtain more detail on the `GroupState.UNKNOWN` state. I'm not sure whether I like this, but included for completeness.
1) In AK 5.0, complete the tidying up by removing `ConsumerGroupState`, `ConsumerGroupListing` and `ListConsumerGroupsOptions`.

@chia7712 @dajac wdyt?

Reply:
~~The idea is interesting to me. 

Unifying group state can eliminate many duplicate code. However, the side effect is that users can't recognize the "true" states for different group type from public APIs. For example, users should not use Reconciling to query share groups as it is nonexistent state in share consumer. This concern can be handled by good docs I think.

I'm +1 to this idea, Additionally, the raw string is an orthogonal issue, so we can discuss that later.